### ISSUE TITLE
Error: Unable to access jar file

### ISSUE THREAD

**Issue by codingChewie** [electron-userland/electron-builder#3863]
**^20.39.0**: 

```
"devDependencies": {
    "devtron": "^1.4.0",
    "electron": "4.1.2",
    "electron-builder": "^20.39.0",
    "eslint": "^5.16.0"
}
```

**all**: 

Using Electron Builder I'm trying to include a jar file and it's supporting library. Per research I have found no solution but a similar question was made: "[How to Package External files with Electron](https://discuss.atom.io/t/how-to-package-external-files-with-electron/28036)" In my main.js I can reference this jar with either:

```
const jarjar = __dirname + '/jarFiles/jarajarbinks.jar'
```

or

```
const jarjar = './jarFiles/jarajarbinks.jar'
```

and the application works in development. When I use Electron builder to build the application I get an error when I try to use the app, the error message is:

> Error: Unable to access jarjar ./jarFiles/jarajarbinks.jar

my _package.json_:

```
"scripts": {
    "start": "electron .",
    "mac": "electron-builder -m",
    "win": "electron-builder -w",
    "build": "electron-builder -mwl"
},
"build": {
    "asar": true,
    "appId": "com.codingChewie.foobar",
    "directories": {
        "buildResources": "build/"
    },
    "extraFiles": [
        "jarFiles/jarajarbinks.jar",
        "jarFiles/lib/*"
    ],
    "mac": {
        "category": "public.app-category.productivity",
        "target": "dmg"
    }
}
```

With electron builder how can I get my application built with the needed jar files and what is the correct way to reference the files in _package.json_?

**Comment by dscalzi** [comment]
I'm not sure about extraFiles, I use extraResources. I also package a jar file with my app stored in the `libraries` folder. In my package.json, I include the following.

```json
extraResources: [
    "libraries"
]
```

I then do something like this in code.

```javascript
if(isDev){
    libPath = path.join(process.cwd(), 'libraries', 'java', 'MyJar.jar')
} else {
    if(process.platform === 'darwin'){
        libPath = path.join(process.cwd(),'Contents', 'Resources', 'libraries', 'java', 'MyJar.jar')
    } else {
        libPath = path.join(process.cwd(), 'resources', 'libraries', 'java', 'MyJar.jar')
    }
}
```

Works fine when packaged.

**Comment by codingChewie** [comment]
Thanks, that's a different approach but still doesn't work. I did install `electron-is-dev` then coded in _main,js_:

```
const isDev = require('electron-is-dev')
const path = require('path')

if (isDev) {
  jarjar = path.join(process.cwd(), 'jarFiles', 'jarajarbinks.jar')
} else {
  if (process.platform === 'darwin') {
    jarjar = path.join(process.cwd(),'Contents', 'Resources', 'jarFiles', 'jarajarbinks.jar')
  } else {
    jarjar = path.join(process.cwd(), 'resources', 'jarFiles', 'jarajarbinks.jar')
  }
}
```

and in _package.json_ I've coded:

```
"extraResources": [
    "jarFiles"
],
```

and I can see the jar files with the lib when I right click and go to `app/Show Package Contents/Contents/Resources/jarFiles` but when I run the app I still get:

> Error: Unable to access jarjar ./Contents/Resources/jarFiles/jarajarbinks.jar

**Comment by codingChewie** [comment]
I've figured out what the issues were. There were two I was facing.

1) in my _package.json_ I was using spaces in the `productName`, example:

```
  "productName": "jar jar binks",
```

which interfered with getting the path to run the jar file.

2) When I was declaring `jarjar = path.join(process.cwd(),'Contents', 'Resources', 'jarFiles', 'jarjarbinks.jar')` I was getting an incorrect path with multiple issues. First it was trying to declare the jar file in _app.asar_ and second it was calling `Contents/Resources/` twice.

To get the app to use the jar file with the correct path I had to modify my conditional to:

```
if (isDev) {
  jarjar = path.join(process.cwd(), 'jarFiles', 'jarjarbinks.jar').replace(/(\s)/g, '\\$1')
} else {
  if (process.platform === 'darwin') {
    jarjar = path.join(__dirname.substring(0, __dirname.lastIndexOf('/')), 'jarFiles', 'jarjarbinks.jar').replace(/(\s)/g, '\\$1')
  }
}
```

with the addition of:

```
"extraResources": [
    "jarFiles"
],
```

No current solution or know if this works for Windows yet as I do not have it installed yet.
